### ISSUE TITLE
Feature request: Restriction of freedom to change primary email addresses

### ISSUE THREAD

**Issue by lordsarcastic** [pennersr/django-allauth#2620]
I noticed that changing primary email requires no verification except that the new email has to be verified. This open up a path to vulnerability issues. 
Take a scenario where a user's password has been compromised. A hacker could just login to the user's account, add their email, verify it and then change the primary email to their(the hacker)  own. 
This means in the database of users falls into the hands of hackers, the whole accounts are gone. 

**What should happen instead**
In the event primary email is going to be changed, a confirmation email should be sent to the present primary email for the user to approve this change. 

I was working on a project that needed this package and I noticed this vulnerability. I put together a hack that works for just this purpose. If the maintainers are interested, I could make a PR on a separate branch of this and the code would be improved upon to be stable and thoroughly tested before being made available to users. This PR would include another feature, one which I'll create another issue on later. 

Thanks for the opportunity

**Comment by 9mido** [comment]
@Lord-sarcastic Good find! Something should definitely be done to fix this. Please share your code below and any explanations of how to use it. Also make a PR so it can be hopefully be put into the main package.

**Comment by pennersr** [comment]
To relativize: this issue is labelled as a vulnerability, but the depicted scenario requires a compromised/vulnerable account in the first place. 

>  This means in the database of users falls into the hands of hackers, the whole accounts are gone.

Can you elaborate? Passwords are not stored in the clear in the database.

**Comment by Lord-sarcastic** [comment]
Yeah, that premise is true. I should put it in a different manner by saying:
An authentication library such as this should factor in possible cases of compromised passwords. In the event passwords have been compromised in cases of extremely sensitive accounts or web applications with high security expectations, the final fallback asides 'safety' of passwords in the database should incorporate confirmation of the change of primary email address.

In terms of the word 'vulnerability', I labelled it as 'potential' in the sense that in some cases such as outlined above, it could become a vulnerability. Perhaps the word 'probable' should be used instead.

Users should not be entirely trusted in this case to have sole and exclusive knowledge of their passwords enough to let them change the primary email as they wish without a confirmation. In the event the issue is not a vulnerability, I still think the feature is worth it. I must also add, it should be there because, why not?

**Comment by bodgerbarnett** [comment]
One thing that occurs to me with the idea of forcing the original email address to approve the change of email address is that the user might be changing their email address because they no longer have access to the old one. Adding this feature would mean users in this situation would be helpless, no?

**Comment by Lord-sarcastic** [comment]
It's mostly a case of preference and/or business decision. Access to email address is more or less an important thing to have. To lose your email address is to lose access to your account in some websites such as Instagram. 

As a fallback, other means of password recovery could exist. This could include phone number verification or secret answers to questions.

To allow free change of primary email address would give a hacker free rein to do almost anything and it would be difficult to recover such an account.

**Comment by 9mido** [comment]
@Lord-sarcastic 

I was able to prevent a secondary email address from becoming a primary email address. But I don't know how to send the verification email to the primary email address instead of the one that requested the primary email address change. 

Inside of views.py class EmailView:

```
    def _action_primary(self, request, *args, **kwargs):
        email = request.POST["email"]
        primary_email = EmailAddress.objects.filter(user=request.user, primary=True).values_list('email', flat=True)
        primary_email_address = EmailAddress.objects.get(user=request.user,primary=True)
        if email not in primary_email:
            messages.error(self.request, "You cannot make this email a primary email address")
            return HttpResponseRedirect(self.request.path_info)
```

**Comment by Lord-sarcastic** [comment]
Nice idea there. In my code where I had to implement this, I used this approach:

- Email address to be used as primary email has to be confirmed and verified first before being used as a primary email
- Send confirmation to current primary email address
- Clicking on confirmation link would change primary email to new email.

In this case, no email has to be sent to the new email and verification is kept.

**Comment by 9mido** [comment]
@Lord-sarcastic Thanks. I made a PR for it. 

Can you show us your code on this issue thread or try to make a PR for with that feature?

**Comment by derek-adair** [comment]
To everyone in this issue, i am doing triage for @pennersr.  

-------------------------
I think this should be closed as a discussion and we need to start thinking about supporting *something* like this.  If someone wants this specific thing it should be fairly easy to extend allauth to do such a thing.  

Also... ya it doesn't track as a "vulnerability" per se - but it does *touch on* a valid point of improvement -- alerting users about actions on their accounts.

* Users logging in to new computers (2fa)
* Important Account status changes
* passwords being changed

Again, not convinced a confirmation email is the best thing here but definitely worth starting to think about alerts.

**Comment by pennersr** [comment]
As part of the recently added MFA functionality a new decorator was introduced: `reauthentication_required()` -- which ensures that before the user can proceed, (s)he needs to be (recently) reauthenticated. This could be used to protect making changes to the email address setup as well.

**Comment by derek-adair** [comment]
> As part of the recently added MFA functionality a new decorator was introduced: `reauthentication_required()` -- which ensures that before the user can proceed, (s)he needs to be (recently) reauthenticated. This could be used to protect making changes to the email address setup as well.

This is a different feature than i'm suggesting but would be a fantastic addition.  This would prevent any users who leave themselves logged in from being compromised.  If this was the specific "vulnerability" that was described I misunderstood and this is important.

--------------

I'm more thinking about "alerts" for when something alert-worthy _event_ happens.  New logins (location/datetime+ relevant details), email changed, account status changes(?), custom user defined events??

**Comment by pennersr** [comment]
You need both, re-authentication prevents your colleague from making fundamental changes to your account when you go for a coffee, and the alerts you mention are needed to notify when changes did happen. The latter is filed as a feature request over here: #3411

**Comment by derek-adair** [comment]
Ok so this should probably be closed, ya?  Seems like there is discussion of some very specific use cases and flows that are not worth supporting and will mostly be resolved / made irrelevant w/ planned features.

**Comment by lordsarcastic** [comment]
Hi there, been a while 😊

Is there any reason supporting a forced email confirmation is not needed per se?

**Comment by derek-adair** [comment]
>Django Allauth Integrated set of Django applications addressing authentication, registration, account management as well as 3rd party (social) account authentication.

I'd file alerts under account management.  I am dispassionate about how it would work just that as a user it would be nice to have.

**Comment by pennersr** [comment]
@lordsarcastic 

> Is there any reason supporting a forced email confirmation is not needed per se?

While this ticket is specifically about changing the primary email, I think we need to zoom out and provide a consistent flow for any important change to the account. For example, it does not make sense to protect the primary email change this way while still allowing someone to turn on MFA, add additional email addresses and so on, using a different flow. And, protecting everything with a confirmation email may be questionable. UX wise, email confirmation flows are asynchronous and do interrupt the user's workflow.  So, while that flow is still pending, do you show anything in the UI, like current config: X, pending config: Y?

So, taking a step back, what you want is to reaffirm that the user making the change does in fact have full control over the account. That to me is a form of re-authentication, which is currently implemented (`@reauthentication_required`) in a way where we ask the user for the password, but that could very well be extended with additional methods, like entering a 2FA code, or, sending some numeric code to the email or phone of the user and letting the user type that in.

**Comment by pennersr** [comment]
> Alerts probably seem like a stretch to implement given that authentication stuff is the focus of the library.

Accounts are needed for authentication purposes, and I regard any flow related towards account management that does have a de facto way of handling things in scope, as it saves projects integrating allauth from doing this over and over again. Now, sending an email containing something like "Your password was changed by IP 1.2.3.4 using browser Mozilla 5.0 (Gecko Firefox)" is common enough to be offered out of the box (using an optional on/off switch).

**Comment by spygency1** [comment]
NEED URGENT HELP?!!

NOTE:MESSAGE ONLY THIS ACCOUNT
#🇬🇧Expert Here To Assist In Any Cyber Issues General_Hacking_Recovery_Spy_Blackmail_Stolen,Lost BTC_Hacked Phones ☎️ 📥
🌎Facebook hacking - recovery
🌎Instagram hacking - recovery
🌎Snapchat hacking - recovery
🌎Lost money(investment scams) recovery
🌎Cheating partner spy (online)
🌎Crypto wallet hacking
🌎Phone hacks📱
🌎Message hacks & tweaks

For swift response send a message to my social media accounts, we are active 24/7 to solve all your Online ETHICAL HACKING issues 

INSTAGRAM 📥 spygency.1
EMAIL 📧 spygency@gmail.com

**Comment by pennersr** [comment]
Closed via d7d8d2f2 -- `ACCOUNT_REAUTHENTICATION_REQUIRED` can be set to `True`.

**Comment by pennersr** [comment]
https://codeberg.org/allauth/django-allauth/issues/2620
